File: s20182019/hw4/hw4.py
import numpy as np
from numpy.linalg import inv

T = 300.0   # Temperature (K)
kB = 1.38066E-23    # Boltzmann constant (J/K)
N_acc = 1.0E+24  # Dopant density (# / cm^3)
Ni = 1.075E+16  # Intrinsic density
q = 1.602192E-19 # Elementary charge (C)

eps0 = 8.85E-12 # Vacuum Permitivity
eps1 = 11.7 # Si permitivity (relative)
eps2 = 3.9 # SiO2 permitivity (relative)

dx = 0.1e-9 # 0.1 nm Spacing
N = 61  # 6 nm thickness

mid1 = 5;   # interface at x=0.5 nm
mid2 = 55;  # interface at x=5.5 nm

A = np.zeros(shape=(N,N),dtype=np.float64)
A[0][0] = 1.0

for i in range(1,N-1):
    if i < mid1:
        A[i][i-1] = eps2
        A[i][i] = -2.0*eps2
        A[i][i+1] = eps2
    elif i is mid1:
        A[i][i-1] = eps2
        A[i][i] = -(eps1+eps2)
        A[i][i+1] = eps1
    elif i < mid2:
        A[i][i-1] = eps1
        A[i][i] = -2.0*eps1
        A[i][i+1] = eps1
    elif i is mid2:
        A[i][i-1] = eps1
        A[i][i] = -(eps1+eps2)
        A[i][i+1] = eps2
    else:
        A[i][i-1] = eps2
        A[i][i] = -2.0*eps2
        A[i][i+1] = eps2
A[N-1][N-1] = 1.0

A_inv = inv(A)
A_inv1 = inv(A)

b = np.zeros(shape=(N),dtype=np.float64)
bf = np.zeros(shape=(N),dtype=np.float64)
for i in range(11):
    b[0] = 0.33374-0.1*i
    b[N-1] = b[0]

    for j in range(mid1,mid2+1):
        if j is mid1:
            b[j] = dx*dx*q*N_acc*0.5/eps0
        elif j is mid2:
            b[j] = dx*dx*q*N_acc*0.5/eps0
        else:
            b[j] = dx*dx*q*N_acc/eps0

    bf[0] = 0.33374-0.1*i
    bf[N-1] = bf[0]



    x = np.zeros(shape=(N),dtype=np.float64)

    x = np.dot(A_inv,b) # Initial potential


    E_d = np.zeros(shape=(N),dtype=np.float64) # Electron density

    for j in range(mid1,mid2+1):
        E_d[j] = Ni*np.exp(q*x[j]/(kB*T))
        if j is mid1:
            bf[j] = dx*dx*q*(N_acc+E_d[j])*0.5/eps0
        elif i is mid2:
            bf[j] = dx*dx*q*(N_acc+E_d[j])*0.5/eps0
        else:
            bf[j] = dx*dx*q*(N_acc+E_d[j])/eps0

    xf = np.zeros(shape=(N),dtype=np.float64)
    xf = np.dot(A_inv,bf)

    f = open('data'+str(i)+'.dat','w')
    for j in range(N):
        f.write("{}\t{}\t{}\t{}\n".format(float(j*0.1),x[j],xf[j],E_d[j]))
    f.close()

File: s20181018/HW4.m
close all;
clear all;

q = 1.602192e-19; % Elementary charge (C)
eps0 = 8.854187817e-12; % Vacuum permittivity, (F/m)
k_B = 1.380662e-23; % Boltzmann constant (J/K) 
T = 300; % Temperature (K)
Delta_x = 0.1e-9; % 0.1 nm spacing
N = 61; % 6 nm thick
x = Delta_x*transpose([0:N-1]); % real space (m)
Inter1 = 6; % At x = 0.5 nm
Inter2 = 56; % At x = 5.5 nm
eps_si = 11.7; % Silicon relative permittivity
eps_ox = 3.9; % Silicon dioxide relative permittivity
Nacc = 1e24; % 1e18/cm^3
ni = 1.075e16; % 1.075e10/cm^3
Voltage_sweep = 0:0.1:1; % Votage step (V)

for i = 1:11;
    gate_potential = 0.33374-Voltage_sweep(i);
    A = zeros(N,N);
    A(1,1) = 1; A(N,N) = 1;
    for j = 2:N-1
        if     (j < Inter1)   A(j,j-1) = eps_ox;    A(j,j) = -2*eps_ox;         A(j,j+1) = eps_ox;
        elseif (j == Inter1)  A(j,j-1) = eps_ox;    A(j,j) = -eps_ox-eps_si;    A(j,j+1) = eps_si;
        elseif (j < Inter2)   A(j,j-1) = eps_si;    A(j,j) = -2*eps_si;         A(j,j+1) = eps_si;
        elseif (j == Inter2)  A(j,j-1) = eps_si;    A(j,j) = -eps_si-eps_ox;    A(j,j+1) = eps_ox;
        elseif (j > Inter2)   A(j,j-1) = eps_ox;    A(j,j) = -2*eps_ox;         A(j,j+1) = eps_ox;
        end
    end
    
    % Boundary value of the electrostatic potential %
    B = zeros(N,1);
    B(1,1) = gate_potential; B(N,1) = gate_potential;
    for j = Inter1:Inter2
        if     (j == Inter1)  B(j,1) = Delta_x*Delta_x*q*Nacc/eps0*0.5;
        elseif (j == Inter2)  B(j,1) = Delta_x*Delta_x*q*Nacc/eps0*0.5;
        else                  B(j,1) = Delta_x*Delta_x*q*Nacc/eps0;
        end
    end
    phi = inv(A)*B;

    % Electron density %
    n = zeros(N,1);
    for j = Inter1:Inter2
        n(j,1) = ni*exp(q*phi(j,1)/(k_B*T));
    end
    
    % Re-calculation of potential by using calculated carrier density %
    B_re = zeros(N,1);
    B_re(1,1) = gate_potential; B_re(N,1) = gate_potential;
    for j = Inter1:Inter2
        if     (j == Inter1)  B_re(j,1) = Delta_x*Delta_x*q*(Nacc+n(j,1))/eps0*0.5;
        elseif (j == Inter2)  B_re(j,1) = Delta_x*Delta_x*q*(Nacc+n(j,1))/eps0*0.5;
        else                  B_re(j,1) = Delta_x*Delta_x*q*(Nacc+n(j,1))/eps0;
        end
    end
    
    phi_re = inv(A)*B_re;
    diff(:,i) = phi-phi_re;
    
    plot(x/1e-9,diff);
    xlabel('Distance (nm)');
    ylabel('Potential (V)');
    legend('0 V','0.1 V','0.2 V','0.3 V','0.4 V','0.5 V','0.6 V','0.7 V','0.8 V','0.9 V','1.0 V');
end

% Initial Condition (i = 1) %
% figure(1);
% plot(x/1e-9,phi(:,i));
% xlabel('Distance (nm)');
% ylabel('Potential (V)');
% figure(2);
% plot(x/1e-9,n(:,i)*1e-6);
% xlabel('Position (nm)');
% ylabel('Electron density (cm^-^3)'); 
% figure(3);
% plot(x/1e-9,phi_re(:,i));
% xlabel('Distance (nm)');
% ylabel('Potential (V)');
% figure(4);
% plot(x/1e-9,diff(:,i));
% xlabel('Distance (nm)');
% ylabel('Potential (V)');
